### PR TITLE
Feature/issue 23 Adding transcript formatters

### PR DESCRIPTION
**Pull request by crhowell**
I have been interacting with this API quite a bit recently, I came across issue #23  and since it seemed like it would be a neat feature to have an I might have some future use cases for it. Thought I would help contribute.

I added much of the implementation y'all discussed for possible implementations posted inside issue #23 and I added a few more features you may like as well.

A few notes for the owner: @jdepoix 

**NOTE 1**:  I **have not** written tests for my formatters just yet but the code I adapted in **should pass all your existing tests**. I did have to modify a few mocks to accept the `format=` kwarg so they would pass. But if you find this PR to your liking, it shouldn't be too difficult for me to add in some additional unittests.

**NOTE 2**: I did notice the `setup.py` says it supports Python version 2, though much of the code looks closer to Python 3 and if it is meant to support Python 2.x there may be some missing syntax. Though I have not attempted to run this project on Python 2 but something I noticed that I thought I would mention while I am here.

<details>
  <summary>Expand this for more info and use-cases</summary>

Using the **CLI**
```
# Example 1: these two lines would be equivalent,
#  json would be the default if not given.

youtube_transcript_api <first_video_id>
youtube_transcript_api <first_video_id> --format json

# Example 2: Changing to another format outputting contents into a file
youtube_transcript_api <first_video_id> --format srt > some_file.srt

# Example 3: If for some reason you need to write multiple transcripts
#  out to a file, though im not sure when I would use this specifically
#  though it is possible.
youtube_transcript_api <first_video_id> <second_video_id> --format json > some_file.json
```

Using the **API class method** call in a script.
```
# These two lines would be equivalent, default is 'json'
YouTubeTranscriptApi.get_transcript(video_id)
YouTubeTranscriptApi.get_transcript(video_id, format='json')
```

Using the **CLI** with a Custom Formatter you can make
> Replacing `custom` with a name you gave it.

```
# Cannot use your own for the CLI at the moment.
youtube_transcript_api <first_video_id> --format custom
```

Using the **API method call** with your Custom Formatter
> Replacing `'custom'` with a name you gave it.

```python
YouTubeTranscriptApi.get_transcript(video_id, format='custom')
```

**Usage**: To build your own Custom Formatter for your script to use.
> NOTE: Wont work for the CLI because of how the module loads, I am open to ideas on how I might make this work.

```python
# file: some_custom_script.py
# ... your other imports

# This is the Abstract Base class for a TranscriptFormatter you need to inherit.
from youtube_transcript_api.formatters import TranscriptFormatter

# formats is a "factory" instances containing Formatter classes.
from youtube_transcript_api.formatters import formats

# Create your Custom Formatter class
class CustomTranscriptFormatter(TranscriptFormatter):
    @classmethod
    def format(cls, transcript_data):
        # ... your custom processing of plain_data ...
        return ['Custom Output Text']
```
> NOTE: Due to a few edge cases the only happy medium I could find for making 
> the existing tests pass and allowing the output to work for a single transcript
> and many transcripts. Your `return` value of `.format()` should ideally be:
> a List containing strings, where each string is a "stringified" version of an
> entire transcript.
> To keep consistency

> For a single transcript:
> Ex, output: ['transcript 1\nline1\nline2']

> For multiple:
> Ex, output: ['transcript 1\nline1\nline2', 'transcript 2\nline1\line2']

And then continuing on further in your script.

```python
# Add your custom class to the factory instance, giving it a unique name.
# this wont prevent you from replacing/overwriting existing formatter name->classes.
formats.add_formatter('custom_name', CustomTranscriptFormatter)

# Now it is able to be retrieved from the factory instance by the name you gave it.
transcript = YouTubeTranscriptApi.get_transcript(<video_id>, format='custom_name')
```

> NOTE: this means `transcript` will be a string but if formatting was done properly,
>  if you are working with JSON for example, `json.loads(transcript)` should give you the
>  data in proper python data types that you can work with. The formatter output had to be made
>  consistent or I would of had to rewrite the API/CLI implementations completely or
>  do entirely too much editing of existing unittests. My goal was to leave the current
>  tests in-tact.
</details>

### CODE REVIEW

**Comment by jdepoix** [comment]
Hi @crhowell,
thanks a lot for your contribution! 😊 👍 
I should have some time to have a look at it tomorrow to give you some feedback.
Regarding Python 2: The travis build runs the test suite in the following python versions 2.7, 3.4, 3.5, 3.6. So that's one of the reasons why keeping the test coverage up is important. But you can add the tests once I gave you feedback, so you won't have to refactor them in case you'll have to change something.
Regarding the custom formatters: I don't think that it is necessary to make custom formatters addable via the CLI. If you're a CLI user you're probably not looking to interact with the internals on that level. If you want to something more custom, you'll probably write some code on your own either way. But (not having looked at the code yet) I don't think it is necessary to add a feature for adding custom formatters anyways. If you want to write a custom formatter, you just take whatever the API returns and then format it. No reason to make it part of the module. But I'll think about that again when I am looking at the code.
Either way, thanks a lot for your efforts!

**Comment by crhowell** [comment]
Hi @jdepoix 

Thanks for touching on a few of the notes I left. I am not that well versed in Travis yet, so some of those parts will be a learning experience for me. I will keep Python v2 in mind when I go back through my PR since I do have a habit of writing Python 3.6+ code from working in Django 2+ so much. 

Again this PR was to address the discussion yall were having. Mainly to give an extra --format arg in the CLI but also give the ability from a custom script to pass a format kwarg. But it can always be refactored to work however it needed. I am open to ideas for the code I am adding so when you do get the time I will be happy to hear your thoughts. 😃 

Normally I would just have written in tests to keep coverage up, but my daily workload is intense right now so I definitely avoided taking a test driven approach for the sake of time. But it is something I can/will add later.

**Comment by crhowell** [comment]
Hi @jdepoix I've added tests to cover all the missings lines `coverage report` listed out for formatters.py. 

Since the first commit on this PR, it has been showing Lines 5-7 of _api.py needing tests. I recall your coverage was at 100% though I never modified those lines. I am confused how this wasnt failing checks before but is now?

I was going to add a `# pragma: no cover` but wasn't sure if that is the route you wanted to take for those few lines handling imports? 👀

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [inline youtube_transcript_api/formatters.py:120]: As explained in the other comment, I don't think that this is needed.

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [inline youtube_transcript_api/formatters.py:26]: Why are you not using the `datetime` module?

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [inline youtube_transcript_api/formatters.py:103]: I don't know much about srt, so I'll just trust you on this 😄

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [inline youtube_transcript_api/formatters.py:68]: I guess you can remove `Transcript` from the name of all those formatter implementations. Feels a bit redundant.

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [inline youtube_transcript_api/formatters.py:78]: This `format` method returns a python dict not json, while the `combine` method actually returns a json string. This is not really the expected behaviour I'd say. It also is inconsistent to how the `SRTTranscriptFormatter` works.
However, calling `format` first and then using the results in `combine` would require you to deserialize the json again (if both methods would return a json string), which kinda confirms my feeling that the `combine` method is not an ideal solution.
I think I would prefer having a `format_transcript` and `format_transcripts` method, with both receiving unformatted data. You could still provide a default implementation of `format_transcripts` like:
```pyhton
self.DELIMITER.join(self.format(transcript) for transcript in transcripts)
```
but the class would be more flexible. 
Don't forget adjusting the CLI to this, tho 😊

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [inline youtube_transcript_api/formatters.py:174]: I wouldn't want this feature to change the default behaviour of this module in case no format is provided by the user. Therefore, I'd suggest also adding a `PrettyPrintFormatter` which implements the current default behaviour.

File: youtube_transcript_api/test/test_api.py
Comment (jdepoix) [inline youtube_transcript_api/test/test_api.py:164]: all those adjustments to these tests won't be needed when `format` is removed from `get_transcript`

File: youtube_transcript_api/test/test_formatters.py
Comment (jdepoix) [inline youtube_transcript_api/test/test_formatters.py:15]: I'd prefer having individual test cases for each formatters. I think this would make it cleaner and more extensible, when other formatters are added in the future. You could have then share the same base class if you want them to use the same setup method.

**Comment by crhowell** [comment]
> There are a few design decisions I made when I initially coded this, which I regret in hindsight, but I didn't give them too much thought at the time, as I never expected this to become something which actually gets used by other people

For that quote above, totally no judgement from me on your decisions made. I have been there and I can totally empathize with any reasoning you have for the road you chose at that time. And also, dont all great projects start that way? Someone wrote code for their own use not thinking it would be something others might want too? 😄 

My mindset going in to doing this PR was that:

1. Ideally I want to immitate the scheme/style of how your code looks and operates under the hood so it looks and others can interact with it in the same fashion and feels like it belongs there.
2. I wanted to try to get as close as I could to formatting the data at the soonest available time rather than later. Not that performance is a worry at this time. But in hindsight this is definitely where things can break and probably why I was fighting some of the code so much on both CLI and API sides since I am impacting return values.

I **really appreciate** you breaking down your thoughts for understanding, this totally helps gauge where your mind is at. 💯 

And my take away from this discussion, the way you envision the interface for formatters would be more like:

**Let me know if this accurately reflects your thoughts:**
Once the final returned result comes back from the series of Transcript (_transcripts.py) related calls that happen, the formatters would happen after the fact. Basically some post processing that would be applied or wrapped around that result to return a new result of that given format? And this would be based on when a formatter arg is specified for the CLI and the API wouldn't even reference it because for custom scripts that person would just need to know about importing the formatters and directly applying the formatter to their result within their own script. So documentation will definitely help for that.

If I am understanding all that ☝️  correctly, I am definitely open to refactoring ALL of my submitted code up to this point to fit that. I am not worried about any time lost. This has been fun and has been a learning experience for me which I always love.

File: youtube_transcript_api/_cli.py
Comment (crhowell) [reply youtube_transcript_api/_cli.py:28]: In regards to the `parsed_args.format` being `None` this would fail if I had not chosen to use an underlying `defaultdict` with `JSONTranscriptFormatter` as the default for the TranscriptFormatterFactory `._formatters` attribute. I kind of hid this away but because it is a defaultdict using bracket notation such as this line:

```python
formats = TranscriptFormatterFactory()
formats.get_formatter(None) # <-- should default return to JSONTranscriptFormatter
```
So it shouldn't cause an error. I kept fighting myself on whether I wanted to just pass `'json'` as the default or just use `None` because either way I was still going to have to specify a default action to take for a Formatter or raise an Exception. But the initial discussion was to default to JSON so that was my thinking there.

File: youtube_transcript_api/_cli.py
Comment (crhowell) [reply youtube_transcript_api/_cli.py:40]: I felt the same way when I wrote it. 🙈 It felt like a hacky attempt to make the CLI and API play nice between 1 video and 2 or more videos and because JSON strings were working a little different than regular strings on different payloads, especially if you want to be able to `json.loads`.

File: youtube_transcript_api/formatters.py
Comment (crhowell) [reply youtube_transcript_api/formatters.py:29]: Yes! This was actually the first thing I seen that made me question if the code was actually supporting Python 2. But I dont believe you were doing anything that was requiring true Python objects as a base class up to this point. Or at least Python wasnt complaining about it! 😂

File: youtube_transcript_api/formatters.py
Comment (crhowell) [reply youtube_transcript_api/formatters.py:70]: The base class defines DELIMITER as the thing you want to join all your data on, the `.combine` method uses it.

```python
   @classmethod
    def combine(cls, transcripts):
        return cls.DELIMITER.join(
                str(transcript) for transcript in transcripts)
```

Again this felt like a hack just to satisfy potentially different types of data. Most of the formatters are strings. But JSON data itself was kind of unique when you wanted to convert it to stringified JSON with `json.dumps`. Which you have to do with the entire structure. But if you are working with 1 transcript at a time and appending them to a collection later (which is what the CLI does) then you need some way to "combine" them all back into each other to make a proper JSON string that you could call `json.loads` on without failing.  This is where I fought the CLI and API the most with return values. Hindsight is 20/20 😅

File: youtube_transcript_api/formatters.py
Comment (crhowell) [reply youtube_transcript_api/formatters.py:174]: The current behavior for this is hidden away behind that `defaultdict` 

```python
# because -> self._formatters = defaultdict(JSONTranscriptFormatter)

def get_formatter(self, name):
        return self._formatters[name]  # if name is None, JSONTranscriptFormatter should be returned.
```

I could totally make this default instead to `PrettyPrintFormatter` if that is preferred instead. So then do you still want the default for CLI to be json? If so, then it would make sense to change the default `parsed_args.format` to no longer be `None` and instead be `'json'` and in a custom script a user would then have to specify `format='json'` explicitly.

File: youtube_transcript_api/formatters.py
Comment (crhowell) [reply youtube_transcript_api/formatters.py:35]: Ahh, you might be right. This was likely a left-over artifact from my initial idea I had for implementation. 

So my base class was originally this idea of a "parser" for changing the formatting instead of formatter. So that I could parse the incoming "plain data" before it was ever converted to its List<Dict> format by **parser code you have in place below**. So we could skip that initial processing and allow the user to customize that data earlier on.

I got this working, but would of have to overhaul at least all of the current tests you had. But it helped going from that right into making it "Formatters" instead. 


```python
class _TranscriptParser():
    HTML_TAG_REGEX = re.compile(r'<[^>]*>', re.IGNORECASE)

    def parse(self, plain_data):
        # ... code... 
```

File: youtube_transcript_api/formatters.py
Comment (crhowell) [reply youtube_transcript_api/formatters.py:39]: I agree. I was actually rethinking formatters the other day before writing tests. Rethinking the same thing, a formatter as an instance might be much more flexible/powerful. Thinking about them in similar terms of how a Django REST Framework Serializer works. 👍   How you can pass a Model instance into a Serializer and it can translate that into a given format.

File: youtube_transcript_api/_cli.py
Comment (crhowell) [reply youtube_transcript_api/_cli.py:38]: Should be offloaded to the DELIMITER and `.combine` method for the transcripts so they can be customized for a given format. I think I had YAML format on the mind when writing it. To separate multiple "documents" in YAML you would need to separate each document with a `---` instead of newline breaks. Was trying to leave it up to the Formatter to determine desired output basically without hardcoding in newlines which might not be preferred for a specific format.

But I am just realizing exception messages are not accounted for so they will likely not look nice at all if displayed with more than 1 exception or an exception mixed with the output of a transcript. 🤦

File: youtube_transcript_api/formatters.py
Comment (crhowell) [reply youtube_transcript_api/formatters.py:26]: Good question! I didn't look into `datetime` yet but will definitely. I re-used an implementation seen in issue #23 from this person who displayed the start of what a formatter might look like:

https://github.com/jdepoix/youtube-transcript-api/issues/23#issuecomment-618953435

File: youtube_transcript_api/formatters.py
Comment (crhowell) [reply youtube_transcript_api/formatters.py:103]: I am still learning about the SRT specs. So dont trust it! 🙈 

Again I used https://github.com/jdepoix/youtube-transcript-api/issues/23#issuecomment-618953435 PR as a reference for the implementation. It "appears" to be correct, but yesterday I found a bug with the actual calculated times having overlap which I dont believe is right. 

So this will be getting another pass before we merge.

File: youtube_transcript_api/_cli.py
Comment (jdepoix) [reply youtube_transcript_api/_cli.py:28]: Ah yeah, didn't notice that.
However, the pprinted dict is the default in the current version, not JSON. I'd like to keep it that way, to not break the API more than necessary. Therefore, you'd have to also implement a `PrettyPrintFormatter`, as already pointed out in the other comment, and then make it the default. Passing around `'json'` won't be an issue, if we don't make the formatters part of `get_transcript`.

File: youtube_transcript_api/_cli.py
Comment (jdepoix) [reply youtube_transcript_api/_cli.py:40]: yeah, I can definitely understand how you got there 😄

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [reply youtube_transcript_api/formatters.py:29]: Yeah, it's not really an issue with code as it is currently. However, it's "correcter" to do so if I want to properly support Python 2. I'll definitely add that!

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [reply youtube_transcript_api/formatters.py:70]: yeah, but you overwrote `combine` of the `JSONTranscriptFormatter` and the implementation doesn't use the `DELIMITER`.

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [reply youtube_transcript_api/formatters.py:39]: For sure, we could implement this using a more generic approach like DRF serializers, but I feel like that would be a bit over-engineered, for this project.

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [reply youtube_transcript_api/formatters.py:26]: Ah, okay. I am pretty sure that this can be done with `datetime`, which probably would be a bit more intuitive to read and less error prone, especially given how common `datetime` is in the python community.

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [reply youtube_transcript_api/formatters.py:103]: Oh, okay. Then we should probably invest some more time to check if this is actually correct. How about adding a srt file in the `test/asset` folder and asserting that the result returned by the formatter is correct?

File: youtube_transcript_api/formatters.py
Comment (jdepoix) [reply youtube_transcript_api/formatters.py:174]: As commented above, the current default formatting is the pprint, not json.

**Comment by jdepoix** [comment]
@crhowell yes, your summary sums it up nicely! 😊 
Besides having a better separation of concerns, the formatter classes could also be used when using `list_transcripts` and fetching transcripts in different languages, when they are treated as a separate processing step and not as a part of `get_transcript`.

File: youtube_transcript_api/formatters.py
Comment (crhowell) [reply youtube_transcript_api/formatters.py:70]: Ahh, good catch!

**Comment by jdepoix** [comment]
Hi @crhowell,
How are things going? No pressure, just wanted to check if you're still planning to finish this at some point, or if I should take over.

**Comment by crhowell** [comment]
> Hi @crhowell,
> How are things going? No pressure, just wanted to check if you're still planning to finish this at some point, or if I should take over.

Hey @jdepoix  things ended up getting **very** busy for me this past month though it has slown down quite a bit since this past Friday. I should have a fair bit of free time, but I also now have some pressing issues around my house to work on such as a plumbing leak in my shower before I hop back into coding mode. 

I will not be offended at all if you took over this issue. 👍  

I definitely wasn't expecting to get as busy as I did when I initially submitted this and was planning to help you implement some code for it. By the end of this week or start of next week I should be fairly free to work on things again. So if you get something implemented before then I can help you test it or find a new issue to help contribute to.

**Comment by jdepoix** [comment]
No worries, as I said there's no pressure, no deadline pressing ;)
Just wanted to check if you're still planning to do this at some point. I am quite busy myself atm, so it'll take me some time until I can free up some time for this. So I guess we'll kinda put this on hold for a bit and see who'll find some time first to take care of driving things forward.

**Comment by crhowell** [comment]
@jdepoix Since I wont be keeping most of these changes from this PR. I would like to close/archive this PR and I can just start a new PR once I have a code implementation of what we discussed in this PR of how the transcript formatters should ideally be applied.

Also because I need to delete my forked copy and re-fork again. For whatever reason I seemed to have accidentally merged my **feature/ISSUE-23** into my own forked `master`. 🙃 Which I think to fix that id have to rollback the head and re-commit that which id rather avoid. 

Though I would like to preserve our discussions we had within this PR since they will be useful to reference as I start working on that formatters code again in the next week.

**Comment by crhowell** [comment]
Hey @jdepoix I was working out some code today. I want to run some thoughts by you that I am having while writing this code or the way its beginning to look so far. 

So my interpretation of formatters is this: The main intent of a Formatter is to receive a transcript (List of Dicts, or Dict containing multiple List of Dicts for many transcripts) and output a String representation of its specified format so that it can be quickly/easily placed into a `file.write()` or `print()` statement within someones custom script? Seeing as `.write()` and `print()` require their arg to be a string type. 

Or do you see more usecases than just that? If we are thinking the same thing here, then I have a few potential implementations I could go with. I think I personally prefer **Scenario 3** but would totally like to hear your thoughts as well before I get to far along. 😃 

<details>
  <summary>SCENARIO 1</summary>

```python

# I do my API request, get a single transcript
transcript = YouTubeTranscriptApi.get_transcript('<video_id>')

# I can pass the transcript to a given formatter...
json_fmt = JSONFormatter(transcript)  # this is a JSONFormatter object, not a string yet

# then I want to write to a file in that given format
with open('my_transcript.json', 'w', encoding='utf-8') as file:
    # Would you prefer this way to convert to a string?
    file.write( str(json_fmt) )
```
</details>

<details>
  <summary>SCENARIO 2</summary>

```python
# ... INSERT repeating lines of code from Scenario 1....

with open('my_transcript.json', 'w', encoding='utf-8') as file:
    # or this way, as an instance method that returns a string representation of the format.
    file.write( json_fmt.format() )
```
</details>

<details>
  <summary>SCENARIO 3</summary>

```python
# ... INSERT repeating lines of code from Scenario 1....

with open('my_transcript.json', 'w', encoding='utf-8') as file:
    # or maybe this way?
    # Like how requests library does as a property/calculated property of the Request object.
    file.write( json_fmt.content )
```
</details>

**Comment by jdepoix** [comment]
Hi @crhowell,

My thoughts:
Scenario 1: the `str` method is supposed to provide a textual representation of an object. However in this case it does the "actual work" the object is design for, if you know what I mean. Not liking this one.
Scenario 2: Simple, easy and explicit. No hidden shenanigans. I am all about this one.
Scenario 3: I am not a fan of properties actually doing stuff (other than maybe concatenating a string or handling caching), because as the caller of the property you don't know that there actually is something happening behind the curtains. Also similarly to Scenario 1 I don't think the "actual work" an object is supposed to do should happen in a property.

**Comment by crhowell** [comment]
Hey @jdepoix  

I added the NEW formatters code. If you want to have a look. It looks like Travis didn't complain about those import lines this time. 🤔 

A few things to note:
- My `test/test_formatters.py` has some basic initial tests written. Even though coverage was 100% on it. I feel like it definitely could test more specific parts of the module. I just wanted to get something pushed for you to look at since it has been awhile.
- I also included some description and use-case examples this time with the `README.md` 
- Abstracting a base class formatter might be overkill right now. Unless we/others want to interate and add even more provided formatters. This will also help us share logic if formatters module starts to grow more. 
- I including the optional **kwargs param on `format()` thinking this would be useful especially when leveraging built-in modules that also accept optional kwargs so we can forward them on. I included this as the last part of the example in the README.

Let me know what you think. Or if you have input on what could be better. I'd like to add more tests before we merge it. But I will let you be the deciding factor on that.

**Comment by jdepoix** [comment]
Hi @crhowell,
FYI: I am still very busy right now and just can't find time to look at the PR. I hope that I will be able to do so by the end of next week.
However, as always, I do appreciate you taking the time! 👍

**Comment by crhowell** [comment]
> Hi @crhowell,
> FYI: I am still very busy right now and just can't find time to look at the PR. I hope that I will be able to do so by the end of next week.
> However, as always, you taking the time is very much appreciated! 👍

Hey no worries. Take your time. I am in Oregon, we have many fires going on around my area and many homes have burned down. I won't be very active with writing code this week. If I become unresponsive in the coming weeks, feel free to take over this PR as I probably had to evacuate my home.

**Comment by jdepoix** [comment]
Oh my god, so sorry to hear that! I sincerely hope for the best for you and your loved ones!

**Comment by jdepoix** [comment]
Hi @crhowell, 
happy to hear from you! I didn't forget about this PR, I just was really busy writing my master thesis! 😄 
But I just finished, so I was planning to pick this back up again as soon as possible and get some stuff done which has been neglected in the last few months due to the thesis...

**Comment by jdepoix** [comment]
Hi @crhowell,
sorry for taking so long, but I finally got around to having a look at your code. I read through the discussion which we had when initially starting this PR. I noticed that some ideas got lost through time (most notably the use of `format_transcript` + `format_transcripts` and integration into the CLI). To accelerate this PR a bit I implemented that stuff and pushed it to your branch. Go have a look and let me know what you think. From my point of view this looks good to be merged now, what do you think?

**Review by jdepoix — APPROVED** [review]

**Comment by crhowell** [comment]
@jdepoix Yes, I believe some things definitely got lost with time 😅. I do vaguely remember we had a conversation where I ditched the idea of a Formatters loader and trying to make it work nicely with the CLI and the API. I started down the refactoring path with [this](https://github.com/jdepoix/youtube-transcript-api/pull/63#discussion_r452989202) in mind as well as [this](https://github.com/jdepoix/youtube-transcript-api/pull/63#issuecomment-656780708) I posted after that last one. 

So basically running with this idea of formatting using individual instances that run an instance method to produce the output. Similar to the idea of a DRF Serializer but of course much more simplified for our use-case.

Basically, the use-case I implemented was primarily targeting API and not CLI uses. But once this API feature is flushed out it should be pretty easy to just have the CLI logic pass a retrieved transcript right into a given formatter and have it output that stringified format.

The API use-case I was envisioning.

```python
# API: your_custom_script.py

from youtube_transcript_api import YouTubeTranscriptApi
from youtube_transcript_api.formatters import JSONFormatter

# Must be a single transcript.
transcript = YouTubeTranscriptApi.get_transcript('<VIDEO_ID>')

formatter = JSONFormatter()

# .format_transcript(transcript) turns the transcript into a JSON formatted string.
json_formatted = formatter.format_transcript(transcript)


# Now we can write it out to a file.
with open('your_filename.json', 'w', encoding='utf-8') as json_file:
    json_file.write(json_formatted)

# Now should have a new JSON file representation of the transcript.
```

I will grab another ☕  and have a look at what you added. 🙂

**Comment by crhowell** [comment]
Ah, I see what you were thinking. All my initial attempts to make the CLI and API work together. You basically were envisioning my initial logic directly in the CLI to be moved into this FormatterLoader class where it can return back out an instance when asked. And then that individual instance handles the translation.

And of course, you handled the multiple cases of the CLI or API where I might be dealing with multiple video transcripts so a method that handles the iteration but offloads each iteration to the single transcript method. 

I was confused at first why we rolled back to the Loader but I see why you went back to that, I like it. I basically can do this now

```python
# Retrieve a list of possible transcript.
transcript = YouTubeTranscriptApi.get_transcript('VIDEO_ID_HERE')
# re-usable loader
loader = FormatterLoader()
# Ask the Loader for a given formatter, returns appropriate formatter instance.
text_formatter = loader.load('text') # TextFormatter() is returned.
# ask the formatter instance to translate our transcript to its format.
output = text_formatter.format_transcript(transcript)
print(output) # outputs transcript in plain text.
```

**Comment by crhowell** [comment]
@jdepoix I will have to let you resolve this conflict with the README. After the last few PR merges, closing and deleting branches. I ran into an interesting case where after I re-synced my forked copy it was auto-committing in the sync/merge process. Which put my fork at 3 commits ahead.

I decided to delete my forked repo and recreate it fresh. It isn't meant to impact the upstream PR but it seems it did break my ability to resolve conflicts via committing from my forked copy as they aren't synced to each other anymore. 

In hindsight, I wish I would have known about the `git rebase @{u}` command as that probably would have solved my issue when I first had seen them between each re-sync with upstream. 🤦

**Comment by jdepoix** [comment]
@crhowell well, as I did my last changes by pushing to your fork, this also broke my ability to push changes to this PR 😄 

While I am sure there is some way to fix this, I think the easiest way is to just close this PR and open a new one. Since I will merge it straight away I don't mind the conversation being lost in this case.

**Pull request closed without merging**
